### PR TITLE
Sample track fx

### PR DESCRIPTION
**Pull request by PhysSong**
This is a part of LMMS#3632: 
* [x] Sample Tracks can be sent to a FX channel
* [x] The selected FX channel can be saved and loaded into project files
* [ ] `DataFile::upgrade()` *(I think not needed)*
* [x] Move `fxLineLcdSpinBox` class into its own file *(and update its coding style)*
* [x] `FxLineLcdSpinBox` has special features that apply to instrument tracks only
* [x] Once closed using the `X` button and opened again, the controls no longer appear and the window is tiny
* [x] No Volume and Panning controls in the window

Please use "Rebase and merge" if you merge it.

### CODE REVIEW

File: include/FxLineLcdSpinBox.h [hunk 1]
```diff
@@ -27,22 +27,26 @@
...
+class TrackView;
 
 
 class FxLineLcdSpinBox : public LcdSpinBox
 {
 	Q_OBJECT
 public:
-	FxLineLcdSpinBox(int numDigits, QWidget * parent, const QString& name) :
-		LcdSpinBox(numDigits, parent, name)
+	FxLineLcdSpinBox( int numDigits, QWidget * parent, const QString& name, TrackView * tv = NULL ) :
```
Comment (grejppi) [inline include/FxLineLcdSpinBox.h]: Space after opening parenthesis is no longer the rule

Comment (PhysSong) [inline include/FxLineLcdSpinBox.h]: @grejppi Okay. How about other parts?

Comment (grejppi) [inline include/FxLineLcdSpinBox.h]: I'd say to use the new style if it's a new function, new class, or a major rewrite of something 😅

Comment (PhysSong) [inline include/FxLineLcdSpinBox.h]: Understanded.

**Comment by PhysSong** [comment]
@grejppi Done.

File: include/FxLineLcdSpinBox.h [hunk 2]
```diff
@@ -27,22 +27,26 @@
...
 
 
 class FxLineLcdSpinBox : public LcdSpinBox
 {
 	Q_OBJECT
 public:
-	FxLineLcdSpinBox(int numDigits, QWidget * parent, const QString& name) :
-		LcdSpinBox(numDigits, parent, name)
+	FxLineLcdSpinBox(int numDigits, QWidget * parent, const QString& name, TrackView * tv = NULL) :
+		LcdSpinBox(numDigits, parent, name), m_tv( tv )
```
Comment (grejppi) [inline include/FxLineLcdSpinBox.h]: `(tv)` instead of `( tv )`

File: include/SampleTrack.h [hunk 3]
```diff
@@ -198,12 +220,77 @@ public slots:
...
+} ;
+
+
+
+class SampleTrackWindow : public QWidget, public ModelView,
+								public SerializingObjectHook
+{
+	Q_OBJECT
+public:
+	SampleTrackWindow(SampleTrackView * _tv);
```
Comment (grejppi) [inline include/SampleTrack.h]: Parameter names don't have to begin with underscore

File: include/SampleTrack.h [hunk 4]
```diff
@@ -198,12 +220,77 @@ public slots:
...
+	{
+		return castModel<SampleTrack>();
+	}
+
+	const SampleTrack * model() const
+	{
+		return castModel<SampleTrack>();
+	}
+
+	void setSampleTrackView(SampleTrackView * _tv);
```
Comment (grejppi) [inline include/SampleTrack.h]: Same here (underscore)

File: include/SampleTrack.h [hunk 5]
```diff
@@ -198,12 +220,77 @@ public slots:
...
+
+	SampleTrackView *sampleTrackView()
+	{
+		return m_stv;
+	}
+
+
+public slots:
+	void textChanged(const QString & _new_name);
```
Comment (grejppi) [inline include/SampleTrack.h]: `newName` (LMMS uses camel case)

File: include/SampleTrack.h [hunk 6]
```diff
@@ -198,12 +220,77 @@ public slots:
...
+	void toggleVisibility(bool _on);
+	void updateName();
+
+
+protected:
+	// capture close-events for toggling sample-track-button
+	virtual void closeEvent(QCloseEvent * _ce);
```
Comment (grejppi) [inline include/SampleTrack.h]: Underscore

File: include/SampleTrack.h [hunk 7]
```diff
@@ -198,12 +220,77 @@ public slots:
...
+
+	virtual void saveSettings(QDomDocument & _doc, QDomElement & _this);
```
Comment (grejppi) [inline include/SampleTrack.h]: Underscore

File: include/SampleTrack.h [hunk 8]
```diff
@@ -198,12 +220,77 @@ public slots:
...
+	virtual void loadSettings(const QDomElement & _this);
```
Comment (grejppi) [inline include/SampleTrack.h]: Underscore

File: src/core/Track.cpp [hunk 9]
```diff
@@ -1842,13 +1842,15 @@ void TrackOperationsWidget::updateMenu()
 	{
 		toMenu->addAction( tr( "Clear this track" ), this, SLOT( clearTrack() ) );
 	}
-	if( InstrumentTrackView * trackView = dynamic_cast<InstrumentTrackView *>( m_trackView ) )
+	if(QMenu *fxMenu = m_trackView->createFxMenu(tr("FX %1: %2"), tr("Assign to new FX Channel")))
```
Comment (grejppi) [inline src/core/Track.cpp]: Space is used after keywords (`if`, `while`, `for`, etc) before the opening parenthesis
```c
if (QMenu...)
```

File: src/core/Track.cpp [hunk 10]
```diff
@@ -1842,13 +1842,15 @@ void TrackOperationsWidget::updateMenu()
...
 	{
-		QMenu *fxMenu = trackView->createFxMenu( tr( "FX %1: %2" ), tr( "Assign to new FX Channel" ));
 		toMenu->addMenu(fxMenu);
+	}
 
+	if(InstrumentTrackView * trackView = dynamic_cast<InstrumentTrackView *>(m_trackView))
```
Comment (grejppi) [inline src/core/Track.cpp]: Space after `if`

File: src/core/Track.cpp [hunk 11]
```diff
@@ -2597,6 +2599,19 @@ void TrackView::update()
 
 
 
+/*! \brief Create a menu for assigning/creating channels for this track.
+ *
+ */
+QMenu * TrackView::createFxMenu( QString title, QString newFxLabel )
```
Comment (grejppi) [inline src/core/Track.cpp]: Spaces

File: src/gui/widgets/FxLineLcdSpinBox.cpp [hunk 12]
```diff
@@ -50,14 +50,9 @@ void FxLineLcdSpinBox::contextMenuEvent(QContextMenuEvent* event)
 
 	QPointer<CaptionMenu> contextMenu = new CaptionMenu(model()->displayName(), this);
 
-	// This condition is here just as a safety check, fxLineLcdSpinBox is aways
-	// created inside a TabWidget inside an InstrumentTrackWindow
-	// FIXME this won't necessarily be true anymore
-	if (InstrumentTrackWindow* window =
-		dynamic_cast<InstrumentTrackWindow*>((QWidget*)this->parent()->parent()))
+	if ( QMenu *fxMenu = m_tv->createFxMenu(
+				tr( "Assign to:" ), tr( "New FX Channel" ) ) )
```
Comment (grejppi) [inline src/gui/widgets/FxLineLcdSpinBox.cpp]: Spaces

File: src/tracks/SampleTrack.cpp [hunk 13]
```diff
@@ -784,60 +787,331 @@ SampleTrackView::SampleTrackView( SampleTrack * _t, TrackContainerView* tcv ) :
...
+	fxMenu->addSeparator();
+
+	for (int i = 0; i < Engine::fxMixer()->numChannels(); ++i)
+	{
+		FxChannel * currentChannel = Engine::fxMixer()->effectChannel( i );
+
+		if ( currentChannel != fxChannel )
+		{
+			QString label = tr( "FX %1: %2" ).arg( currentChannel->m_channelIndex ).arg( currentChannel->m_name );
+			QAction * action = fxMenu->addAction( label, fxMenuSignalMapper, SLOT( map() ) );
```
Comment (grejppi) [inline src/tracks/SampleTrack.cpp]: Spaces

File: src/tracks/SampleTrack.cpp [hunk 14]
```diff
@@ -784,60 +787,331 @@ SampleTrackView::SampleTrackView( SampleTrack * _t, TrackContainerView* tcv ) :
...
+	fxMenu->addAction( newFxLabel, this, SLOT( createFxLine() ) );
+	fxMenu->addSeparator();
+
+	for (int i = 0; i < Engine::fxMixer()->numChannels(); ++i)
+	{
+		FxChannel * currentChannel = Engine::fxMixer()->effectChannel( i );
+
+		if ( currentChannel != fxChannel )
+		{
+			QString label = tr( "FX %1: %2" ).arg( currentChannel->m_channelIndex ).arg( currentChannel->m_name );
```
Comment (grejppi) [inline src/tracks/SampleTrack.cpp]: Spaces

File: src/tracks/SampleTrack.cpp [hunk 15]
```diff
@@ -784,60 +787,331 @@ SampleTrackView::SampleTrackView( SampleTrack * _t, TrackContainerView* tcv ) :
...
+	QSignalMapper * fxMenuSignalMapper = new QSignalMapper(fxMenu);
+
+	fxMenu->addAction( newFxLabel, this, SLOT( createFxLine() ) );
+	fxMenu->addSeparator();
+
+	for (int i = 0; i < Engine::fxMixer()->numChannels(); ++i)
+	{
+		FxChannel * currentChannel = Engine::fxMixer()->effectChannel( i );
+
+		if ( currentChannel != fxChannel )
```
Comment (grejppi) [inline src/tracks/SampleTrack.cpp]: Spaces

File: src/tracks/SampleTrack.cpp [hunk 16]
```diff
@@ -784,60 +787,331 @@ SampleTrackView::SampleTrackView( SampleTrack * _t, TrackContainerView* tcv ) :
...
+	QMenu *fxMenu = new QMenu( title );
+
+	QSignalMapper * fxMenuSignalMapper = new QSignalMapper(fxMenu);
+
+	fxMenu->addAction( newFxLabel, this, SLOT( createFxLine() ) );
+	fxMenu->addSeparator();
+
+	for (int i = 0; i < Engine::fxMixer()->numChannels(); ++i)
+	{
+		FxChannel * currentChannel = Engine::fxMixer()->effectChannel( i );
```
Comment (grejppi) [inline src/tracks/SampleTrack.cpp]: Spaces

File: src/tracks/SampleTrack.cpp [hunk 17]
```diff
@@ -784,60 +787,331 @@ SampleTrackView::SampleTrackView( SampleTrack * _t, TrackContainerView* tcv ) :
...
+	if ( title.contains( "%2" ) )
+	{
+		title = title.arg( channelIndex ).arg( fxChannel->m_name );
+	}
+
+	QMenu *fxMenu = new QMenu( title );
+
+	QSignalMapper * fxMenuSignalMapper = new QSignalMapper(fxMenu);
+
+	fxMenu->addAction( newFxLabel, this, SLOT( createFxLine() ) );
```
Comment (grejppi) [inline src/tracks/SampleTrack.cpp]: Spaces

File: src/tracks/SampleTrack.cpp [hunk 18]
```diff
@@ -784,60 +787,331 @@ SampleTrackView::SampleTrackView( SampleTrack * _t, TrackContainerView* tcv ) :
...
+
+	FxChannel *fxChannel = Engine::fxMixer()->effectChannel( channelIndex );
+
+	// If title allows interpolation, pass channel index and name
+	if ( title.contains( "%2" ) )
+	{
+		title = title.arg( channelIndex ).arg( fxChannel->m_name );
+	}
+
+	QMenu *fxMenu = new QMenu( title );
```
Comment (grejppi) [inline src/tracks/SampleTrack.cpp]: Spaces

File: src/tracks/SampleTrack.cpp [hunk 19]
```diff
@@ -784,60 +787,331 @@ SampleTrackView::SampleTrackView( SampleTrack * _t, TrackContainerView* tcv ) :
...
 {
-	m_effWindow->deleteLater();
+	int channelIndex = model()->effectChannelModel()->value();
+
+	FxChannel *fxChannel = Engine::fxMixer()->effectChannel( channelIndex );
+
+	// If title allows interpolation, pass channel index and name
+	if ( title.contains( "%2" ) )
+	{
+		title = title.arg( channelIndex ).arg( fxChannel->m_name );
```
Comment (grejppi) [inline src/tracks/SampleTrack.cpp]: Spaces

File: src/tracks/SampleTrack.cpp [hunk 20]
```diff
@@ -784,60 +787,331 @@ SampleTrackView::SampleTrackView( SampleTrack * _t, TrackContainerView* tcv ) :
...
-SampleTrackView::~SampleTrackView()
+QMenu * SampleTrackView::createFxMenu(QString title, QString newFxLabel)
 {
-	m_effWindow->deleteLater();
+	int channelIndex = model()->effectChannelModel()->value();
+
+	FxChannel *fxChannel = Engine::fxMixer()->effectChannel( channelIndex );
+
+	// If title allows interpolation, pass channel index and name
+	if ( title.contains( "%2" ) )
```
Comment (grejppi) [inline src/tracks/SampleTrack.cpp]: Spaces

File: src/tracks/SampleTrack.cpp [hunk 21]
```diff
@@ -784,60 +787,331 @@ SampleTrackView::SampleTrackView( SampleTrack * _t, TrackContainerView* tcv ) :
...
 
 
-
-SampleTrackView::~SampleTrackView()
+QMenu * SampleTrackView::createFxMenu(QString title, QString newFxLabel)
 {
-	m_effWindow->deleteLater();
+	int channelIndex = model()->effectChannelModel()->value();
+
+	FxChannel *fxChannel = Engine::fxMixer()->effectChannel( channelIndex );
```
Comment (grejppi) [inline src/tracks/SampleTrack.cpp]: Spaces

File: src/tracks/SampleTrack.cpp [hunk 22]
```diff
@@ -784,60 +787,331 @@ SampleTrackView::SampleTrackView( SampleTrack * _t, TrackContainerView* tcv ) :
...
+{
+	SampleTrack * st = castModel<SampleTrack>();
+	m_volumeKnob->setModel(&st->m_volumeModel);
+
+	TrackView::modelChanged();
+}
+
+
+
+SampleTrackWindow::SampleTrackWindow(SampleTrackView * _stv) :
```
Comment (grejppi) [inline src/tracks/SampleTrack.cpp]: Underscore

File: src/tracks/SampleTrack.cpp [hunk 23]
```diff
@@ -784,60 +787,331 @@ SampleTrackView::SampleTrackView( SampleTrack * _t, TrackContainerView* tcv ) :
...
+}
+
+
+
+/*! \brief Create and assign a new FX Channel for this track */
+void SampleTrackView::createFxLine()
+{
+	int channelIndex = gui->fxMixerView()->addNewChannel();
+
+	Engine::fxMixer()->effectChannel( channelIndex )->m_name = getTrack()->name();
```
Comment (grejppi) [inline src/tracks/SampleTrack.cpp]: Spaces

File: src/tracks/SampleTrack.cpp [hunk 24]
```diff
@@ -784,60 +787,331 @@ SampleTrackView::SampleTrackView( SampleTrack * _t, TrackContainerView* tcv ) :
...
+	assignFxLine(channelIndex);
+}
+
+
+
+
+/*! \brief Assign a specific FX Channel for this track */
+void SampleTrackView::assignFxLine(int channelIndex)
+{
+	model()->effectChannelModel()->setValue( channelIndex );
```
Comment (grejppi) [inline src/tracks/SampleTrack.cpp]: Spaces

File: src/tracks/SampleTrack.cpp [hunk 25]
```diff
@@ -784,60 +787,331 @@ SampleTrackView::SampleTrackView( SampleTrack * _t, TrackContainerView* tcv ) :
...
+
+	gui->fxMixerView()->setCurrentFxLine( channelIndex );
```
Comment (grejppi) [inline src/tracks/SampleTrack.cpp]: Spaces

**Comment by PhysSong** [comment]
@grejppi It's because I copied some codes from InstrumentTrack.cpp, I'll fix 'em quickly.

**Comment by PhysSong** [comment]
Is there something left? If you merge, please don't use normal merge. Either "Rebase and merge" or "Squash and merge" will be fine.

**Comment by PhysSong** [comment]
@grejppi Could you check my updated PR?

**Comment by grejppi** [comment]
Sorry for the wait! Looking good, merging 🎉

**Pull request merged**